File: src/utils/vector2.ts
/**
 * 2D Vector class for position and movement calculations
 */
export class Vector2 {
  constructor(
    public x: number = 0.0,
    public y: number = 0.0
  ) {}

  toString(): string {
    return `Vector2(${this.x}, ${this.y})`;
  }

  equals(other: Vector2): boolean {
    return Math.abs(this.x - other.x) < 1e-9 && Math.abs(this.y - other.y) < 1e-9;
  }

  add(other: Vector2 | [number, number]): Vector2 {
    if (other instanceof Vector2) {
      return new Vector2(this.x + other.x, this.y + other.y);
    } else if (Array.isArray(other) && other.length === 2) {
      return new Vector2(this.x + other[0], this.y + other[1]);
    }
    throw new Error("Can only add Vector2 to Vector2 or [number, number]");
  }

  subtract(other: Vector2 | [number, number]): Vector2 {
    if (other instanceof Vector2) {
      return new Vector2(this.x - other.x, this.y - other.y);
    } else if (Array.isArray(other) && other.length === 2) {
      return new Vector2(this.x - other[0], this.y - other[1]);
    }
    throw new Error("Can only subtract Vector2 from Vector2 or [number, number]");
  }

  multiply(scalar: number): Vector2 {
    if (typeof scalar === 'number') {
      return new Vector2(this.x * scalar, this.y * scalar);
    }
    throw new Error("Can only multiply Vector2 by scalar");
  }

  divide(scalar: number): Vector2 {
    if (typeof scalar === 'number' && scalar !== 0) {
      return new Vector2(this.x / scalar, this.y / scalar);
    }
    throw new Error("Can only divide Vector2 by non-zero scalar");
  }

  negate(): Vector2 {
    return new Vector2(-this.x, -this.y);
  }

  get(index: number): number {
    if (index === 0) return this.x;
    if (index === 1) return this.y;
    throw new Error("Vector2 index out of range");
  }

  magnitude(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  magnitudeSquared(): number {
    return this.x * this.x + this.y * this.y;
  }

  normalize(): Vector2 {
    const mag = this.magnitude();
    if (mag > 0) {
      return new Vector2(this.x / mag, this.y / mag);
    }
    return new Vector2(0, 0);
  }

  distanceTo(other: Vector2): number {
    return this.subtract(other).magnitude();
  }

  distanceSquaredTo(other: Vector2): number {
    return this.subtract(other).magnitudeSquared();
  }

  dot(other: Vector2): number {
    return this.x * other.x + this.y * other.y;
  }

  cross(other: Vector2): number {
    return this.x * other.y - this.y * other.x;
  }

  lerp(other: Vector2, t: number): Vector2 {
    const clampedT = Math.max(0, Math.min(1, t));
    return this.add(other.subtract(this).multiply(clampedT));
  }

  toTuple(): [number, number] {
    return [this.x, this.y];
  }

  toIntTuple(): [number, number] {
    return [Math.floor(this.x), Math.floor(this.y)];
  }

  copy(): Vector2 {
    return new Vector2(this.x, this.y);
  }

  // Static methods
  static zero(): Vector2 {
    return new Vector2(0, 0);
  }

  static one(): Vector2 {
    return new Vector2(1, 1);
  }

  static up(): Vector2 {
    return new Vector2(0, 1);
  }

  static down(): Vector2 {
    return new Vector2(0, -1);
  }

  static left(): Vector2 {
    return new Vector2(-1, 0);
  }

  static right(): Vector2 {
    return new Vector2(1, 0);
  }

  static fromTuple(tuple: [number, number]): Vector2 {
    return new Vector2(tuple[0], tuple[1]);
  }
}

File: src/utils/vector2.py
# Vector2D Class for 2D Vector Operations

import math
from typing import Tuple


class Vector2:
    """2D Vector class for position and movement calculations"""
    
    def __init__(self, x: float = 0.0, y: float = 0.0):
        self.x = float(x)
        self.y = float(y)
    
    def __str__(self) -> str:
        return f"Vector2({self.x}, {self.y})"
    
    def __repr__(self) -> str:
        return self.__str__()
    
    def __eq__(self, other) -> bool:
        if isinstance(other, Vector2):
            return abs(self.x - other.x) < 1e-9 and abs(self.y - other.y) < 1e-9
        return False
    
    def __add__(self, other) -> 'Vector2':
        if isinstance(other, Vector2):
            return Vector2(self.x + other.x, self.y + other.y)
        elif isinstance(other, (tuple, list)) and len(other) == 2:
            return Vector2(self.x + other[0], self.y + other[1])
        else:
            raise TypeError("Can only add Vector2 to Vector2 or tuple/list of length 2")
    
    def __sub__(self, other) -> 'Vector2':
        if isinstance(other, Vector2):
            return Vector2(self.x - other.x, self.y - other.y)
        elif isinstance(other, (tuple, list)) and len(other) == 2:
            return Vector2(self.x - other[0], self.y - other[1])
        else:
            raise TypeError("Can only subtract Vector2 from Vector2 or tuple/list of length 2")
    
    def __mul__(self, scalar) -> 'Vector2':
        if isinstance(scalar, (int, float)):
            return Vector2(self.x * scalar, self.y * scalar)
        else:
            raise TypeError("Can only multiply Vector2 by scalar")
    
    def __rmul__(self, scalar) -> 'Vector2':
        return self.__mul__(scalar)
    
    def __truediv__(self, scalar) -> 'Vector2':
        if isinstance(scalar, (int, float)) and scalar != 0:
            return Vector2(self.x / scalar, self.y / scalar)
        else:
            raise TypeError("Can only divide Vector2 by non-zero scalar")
    
    def __neg__(self) -> 'Vector2':
        return Vector2(-self.x, -self.y)
    
    def __getitem__(self, index) -> float:
        if index == 0:
            return self.x
        elif index == 1:
            return self.y
        else:
            raise IndexError("Vector2 index out of range")
    
    def magnitude(self) -> float:
        """Get the length of the vector"""
        return math.sqrt(self.x * self.x + self.y * self.y)
    
    def magnitude_squared(self) -> float:
        """Get the squared length of the vector (faster than magnitude)"""
        return self.x * self.x + self.y * self.y
    
    def normalize(self) -> 'Vector2':
        """Get a normalized copy of the vector"""
        mag = self.magnitude()
        if mag > 0:
            return Vector2(self.x / mag, self.y / mag)
        return Vector2(0, 0)
    
    def normalized(self) -> 'Vector2':
        """Alias for normalize()"""
        return self.normalize()
    
    def distance_to(self, other) -> float:
        """Get distance to another vector"""
        if isinstance(other, Vector2):
            return (self - other).magnitude()
        else:
            raise TypeError("Can only calculate distance to Vector2")
    
    def distance_squared_to(self, other) -> float:
        """Get squared distance to another vector (faster)"""
        if isinstance(other, Vector2):
            return (self - other).magnitude_squared()
        else:
            raise TypeError("Can only calculate distance to Vector2")
    
    def dot(self, other) -> float:
        """Calculate dot product with another vector"""
        if isinstance(other, Vector2):
            return self.x * other.x + self.y * other.y
        else:
            raise TypeError("Can only calculate dot product with Vector2")
    
    def cross(self, other) -> float:
        """Calculate 2D cross product (returns scalar)"""
        if isinstance(other, Vector2):
            return self.x * other.y - self.y * other.x
        else:
            raise TypeError("Can only calculate cross product with Vector2")
    
    def lerp(self, other, t: float) -> 'Vector2':
        """Linear interpolation between this vector and another"""
        if isinstance(other, Vector2):
            t = max(0, min(1, t))  # Clamp t between 0 and 1
            return self + (other - self) * t
        else:
            raise TypeError("Can only lerp with Vector2")
    
    def to_tuple(self) -> Tuple[float, float]:
        """Convert to tuple"""
        return (self.x, self.y)
    
    def to_int_tuple(self) -> Tuple[int, int]:
        """Convert to integer tuple"""
        return (int(self.x), int(self.y))
    
    def copy(self) -> 'Vector2':
        """Create a copy of this vector"""
        return Vector2(self.x, self.y)
    
    @classmethod
    def zero(cls) -> 'Vector2':
        """Create a zero vector"""
        return cls(0, 0)
    
    @classmethod
    def one(cls) -> 'Vector2':
        """Create a vector with all components as 1"""
        return cls(1, 1)
    
    @classmethod
    def up(cls) -> 'Vector2':
        """Create a vector pointing up"""
        return cls(0, 1)
    
    @classmethod
    def down(cls) -> 'Vector2':
        """Create a vector pointing down"""
        return cls(0, -1)
    
    @classmethod
    def left(cls) -> 'Vector2':
        """Create a vector pointing left"""
        return cls(-1, 0)
    
    @classmethod
    def right(cls) -> 'Vector2':
        """Create a vector pointing right"""
        return cls(1, 0)
    
    @classmethod
    def from_tuple(cls, t: Tuple[float, float]) -> 'Vector2':
        """Create vector from tuple"""
        return cls(t[0], t[1])